File: week4/java/RotateArray_189.java
class Solution {
    public void rotate(int[] nums, int k) {
        
        int len = nums.length;
        k %= len;
        
        // no need to rotate
        if (k == 0) return;
        
        // cyclic moving approach
        // int count = 0;
        // for (int start = 0; count < len; start++) {
        //     int cur = start;
        //     int prev_val =  nums[cur];
        //     do {
        //         cur = (cur + k) % len;
        //         int tmp = nums[cur];
        //         nums[cur] = prev_val;
        //         prev_val = tmp;
        //         count += 1;
        //     } while (cur != start);
        // }
        
        // rotation approach
        reverse(nums, 0, len - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, len - 1);
    }
    
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int tmp = nums[start];
            nums[start] = nums[end];
            nums[end] = tmp;
            start += 1;
            end -= 1;
        }
    }
    
}

File: week4/python/RotateArray_189.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # 1st approach
#         if nums:
#             k %= len(nums)
#             visited = [False]*len(nums)
#             for i in range(len(nums)):
#                 if not visited[i]:
#                     start = i
#                     tmp = nums[start]
#                     while True:
#                         visited[start] = True
#                         next_i = (start + k) % len(nums)
#                         nums[next_i], tmp = tmp, nums[next_i]

#                         if next_i == i:
#                             break
                    
#                         start = next_i
        
        # 2nd approach
        if nums:
            k %= len(nums)
            
            if k == 0:
                return
            
            nums[:] = nums[::-1]
            nums[:k] = nums[:k][::-1]
            nums[k:] = nums[k:][::-1]
